Sources/CacheAdvance/FileHeader.swift

```diff
@@ -0,0 +1,150 @@
+        self.offsetInFileOfOldestMessage = offsetInFileOfOldestMessage
+        self.offsetInFileAtEndOfNewestMessage = offsetInFileAtEndOfNewestMessage
+    }
+
+    init?(from data: Data) {
+        guard data.count == FileHeader.expectedEndOfHeaderInFile else {
+            return nil
+        }
+
+        let versionData = data.subdata(in: FileHeader.Field.version.rangeOfFieldInFile)
+        let maximumBytesData = data.subdata(in: FileHeader.Field.maximumBytes.rangeOfFieldInFile)
+        let overwritesOldMessagesData = data.subdata(in: FileHeader.Field.overwriteOldMessages.rangeOfFieldInFile)
+        let offsetInFileOfOldestMessageData = data.subdata(in: FileHeader.Field.offsetInFileOfOldestMessage.rangeOfFieldInFile)
+        let offsetInFileAtEndOfNewestMessageData = data.subdata(in: FileHeader.Field.offsetInFileAtEndOfNewestMessage.rangeOfFieldInFile)
+
+        let version = UInt8(versionData)
+        let maximumBytes = Bytes(maximumBytesData)
+        let overwritesOldMessages = Bool(overwritesOldMessagesData)
+        let offsetInFileOfOldestMessage = UInt64(offsetInFileOfOldestMessageData)
+        let offsetInFileAtEndOfNewestMessage = UInt64(offsetInFileAtEndOfNewestMessageData)
+
+        self = FileHeader(
+            version: version,
+            maximumBytes: maximumBytes,
+            overwritesOldMessages: overwritesOldMessages,
+            offsetInFileOfOldestMessage: offsetInFileOfOldestMessage,
+            offsetInFileAtEndOfNewestMessage: offsetInFileAtEndOfNewestMessage)
+    }
+
+    // MARK: Internal
+
+    let version: UInt8
+    let maximumBytes: Bytes
+    let overwritesOldMessages: Bool
+    let offsetInFileOfOldestMessage: UInt64
+    let offsetInFileAtEndOfNewestMessage: UInt64
+
+    var asData: Data {
+        Field.allCases.reduce(Data()) { header, field in
+            header + data(for: field)
+        }
+    }
+
+    /// The expected version of the header.
+    /// Whenever there is a breaking change to the header format, update this value.
+    static let version: UInt8 = 1
+
+    /// Calculates the offset in the file where the header should end.
+    static var expectedEndOfHeaderInFile = Field.endOfHeaderMarker.expectedEndOfFieldInFile
+
+    func data(for field: Field) -> Data {
+        switch field {
+        case .version:
+            return Data(version)
+        case .maximumBytes:
+            return Data(maximumBytes)
+        case .overwriteOldMessages:
+            return Data(overwritesOldMessages)
+        case .offsetInFileOfOldestMessage:
+            return Data(offsetInFileOfOldestMessage)
+        case .offsetInFileAtEndOfNewestMessage:
+            return Data(offsetInFileAtEndOfNewestMessage)
+        case .reservedSpace:
+            return Data(repeating: 0, count: field.storageLength)
+        case .endOfHeaderMarker:
+            return Data()
+        }
+    }
+
+    enum Field: Int, CaseIterable {
+        // Header format:
```

Comment:
In the past I recall discussing with you how one benefit of the original structure was that appends were atomic, since we only had to write one contiguous blob at a time.

By adding `offsetInFileOfOldestMessage` and `offsetInFileAtEndOfNewestMessage` are appends still atomic? Or is that that something we’ve lost in favor of being able to load a file more quickly?

Curious to understand and learn how you evaluated the tradeoff (if there was a tradeoff).

Reply:
@bachand good q! I've added my thinking here (and an explanation re how this works in the new world) to the PR description.

Reply:
Nice. PR description is great.
What happens if we crash between writes one and two, or between writes two and three? Is it possible to end up with a corrupt file?
Asking to learn.

Reply:
@bachand forgive me, but I'm going to point you at the comments in the code. If the comments don't elucidate, then I'll need to make the comments better.

Before each metadata write I have a comment that should explain. Here's where I write the [oldest message metadata](https://github.com/dfed/CacheAdvance/pull/22/files#diff-b149efa94be256e60ad02ca7b8be50d5R99) and the [newest message metadata](https://github.com/dfed/CacheAdvance/pull/22/files#diff-b149efa94be256e60ad02ca7b8be50d5R181). Please let me know if there isn't enough context in the comments!

Since I don't explicitly discuss this in the comments: I do not believe it is possible to end up with a corrupt file if a crash happens between these writes.

Reply:
@bachand forgive me, but I'm going to point you at the comments in the code. If the comments don't elucidate, then I'll need to make the comments better.

Before each metadata write I have a comment that should explain. Here's where I write the [oldest message metadata](https://github.com/dfed/CacheAdvance/pull/22/files#diff-b149efa94be256e60ad02ca7b8be50d5R99) and the [newest message metadata](https://github.com/dfed/CacheAdvance/pull/22/files#diff-b149efa94be256e60ad02ca7b8be50d5R181). Please let me know if there isn't enough context in the comments!

Since I don't explicitly discuss this in the comments: I do not believe it is possible to end up with a corrupt file if a crash happens between these writes.

Reply:
Oh great. Thanks for these pointers to the code comments.

> I do not believe it is possible to end up with a corrupt file if a crash happens between these writes.

👌 I wanted to make sure that this PR had considered those edge cases before delving too deep.